rfcs/2020-04-23-2445-log-schema.md

```diff
@@ -0,0 +1,240 @@
+# RFC 2445 - 2020-04-23 - Log Schema Facilities & Standards
+
+A proposal to establish facilities and standards for managing log event schemas.
+
+## Preface
+
+Before we begin, I want to clarify that the goal is this RFC is not to
+establish yet another log schema. That is a precarious exercise that usually
+doesn't end well. Instead, the goal is to establish facilities and standards
+that help Vector users adopt the log schema of their choice. This takes the
+pressure off of establishing the perfect schema. Additionally, A core principle
+of Vector is to be unopinionated and "meet users where they are". This is
+fundamental to helping users migrate to better standards over time, one of
+Vector's prime use cases.
+
+## Motivation
+
+It's becoming clear to me that we need better facilities and standards around
+Vector's internal log schema. The amount of effort and discussion around our
+log schema is growing to the point where it would benefit from standardization
+(#2272, #2273, #2147, #2269, #2270, #2271, #1150, #2480). And like other
+features in Vector, this one has grown organically over time:
+
+1. Vector initially offered very simple `*_key` options in the relevant
+   components and still does. Ex: the `file` source's `file_key` option.
+2. We then ran into problems trying to coordinate keys when mapping them
+   at the sink level. For example, Vector's `timestamp` key is mapped to
+   `date` in the `datadog_logs` sink.
+3. This issue was further exacerbated by the `splunk_hec` sink, which offers
+   a [`source` and `source_type` field][splunk_source_fields] that describes
+   where the data came from (ex: `source` = `TCP:0.0.0.0:4000`).
+4. To solve for this we [introduced a new global `log_schema.*`][pr_1769] set
+   of options that contained only 3 keys: `log_schema.message_key`,
+   `log_schema.timestamp_key`, `log_schema.host_key`.
+
+In addition to the above, other issues have popped up that would be solved with
+the proper log schema:
+
+1. Internal metadata ([#2415])
+2. Adopting existing schemas (]#2423])
+
+## Proposal
+
+1. Move all component level keys to the global `log_schema` set of options.
+2. Remove all component level `*_key` options in favor of the new `log_schema.*`
+   options.
+3. Define a namespace for metadata, such as `log_schema.meta_key`. This is
+   intended for internal Vector metadata that does not describe the log
+   directly.
+4. If the field does not directly describe the log, or it's origine, then it
+   should go under the new `log_schema.meta_key`.
+5. Setting any of these fields to `""` disables the field.
```

Comment:
Can we note that internally Vector *must* store these as `Option<_>` and not store full empty `String` types.